### ISSUE TITLE
Opt-in for default zones and rules

### ISSUE THREAD

**Issue by oddlama** [thelegy/nixos-nftables-firewall#6]
I'd like to configure the firewalls of my servers by beginning with a clean state. Unfortunately as it is right now, a zone and several rules are applied unconditionally. Specifically I'd like to use `local` instead of `fw`, close the DHCPv6 client port 546 and disable the `nixos-firewall` and `ssh` zones. https://github.com/thelegy/nixos-nftables-firewall/blob/6cb25335de6f1fe0722f02573d0cfbaea4cd7ecf/nftables-zoned.nix#L294-L323

Currently I have to use `lib.mkForce` to override their definition, which in turn interferes with the classical modularization of my configuration, since multiple definitions in my config will now not be merged, or require additional global handling. So I was wondering whether you might consider adding a opt-in `useRecommendedDefaults` option, so that the module becomes assumption-free on its own.

Thanks for the great work❤️

**Comment by thelegy** [comment]
Sorry for the delay.

First of all `lib.mkForce` only sets the priority on your config and multiple definitions with the same priority merge fine, so you probably could implement it that way.

Secondly the rules are enabled by default, but using `networking.nftables.firewall.rules.<name>.enable` you can disable them and just use another name. No `lib.mkForce` requiered here.

I do acknowledge though, that disabling a list of options is very different from swithing the default for all of them together when it comes to including new rules in this repo. I have some changes in mind to tackle this:

---

In regard of the `fw`/`local` zone: That exactly one local zone is needed is not an assumption, it is actually enforced, so the only assumtion here is the name. That is easy to solve: Another option will be added for the name (e.g. `networking.nftables.firewall.localZoneName` / `networking.nftables.firewall.localZone`). The definition of that zone's `localZone` can than leave out the `mkDefault` as that is no longer the intended way of switching the name of the local zone. Maybe an assertion/warning can be added when the `localZoneName`'s `localZone` is overridden.

---

In regard of the rules it is a bit more difficult and I am still not sure about that. I want to let them be included and enabled by default, as locking yourself out from ssh for example is not the first expirience I want new users to have.

I was thinking about something like an `enableRecommendedRules` option, defaulting to `true` that you could override to disable the rules. You could set that option to `false` but then selectively anable the rules, you do want to use from this repo.

That still has some problems: I is still not assumption free, the rules are still defined but disabled, so the names of the rules are already in use and the list of names you need to work around can grow over time.

Just putting all my rules in an `mkIf` is not flexible enough in my opinion. And having an option with a list of "included" rules and then some complex `mkIf` logic also does not sound like a nice interface for the problem.

Currently my best idea is to rename all rules in this repo to include a common prefix, so collisions with user-definied rules (not using this prefix) are out of the question. This combined with the `enableRecommendedRules` idea, might make your usecase convenient and also be flexible enough, to enable/disable the included rules.

---

What do you think about this? I think the local zone part is pretty uncontroversial, but the rules part might not be.

**Comment by oddlama** [comment]
> First of all `lib.mkForce` only sets the priority on your config and multiple definitions with the same priority merge fine, so you probably could implement it that way.

That's what I ended up doing for now. It's more of a philosophical problem I guess, because now I can't separate a part of my config into a distinct module (or even a separate flake) without also carrying the lib.mkForce assumption over. So in the end it just doesn't feel right, but you're right - there's no technical problem.

> Secondly the rules are enabled by default, but using `networking.nftables.firewall.rules.<name>.enable` you can disable them and just use another name. No `lib.mkForce` requiered here.
>
> I do acknowledge though, that disabling a list of options is very different from swithing the default for all of them together when it comes to including new rules in this repo. I have some changes in mind to tackle this:

True. That's exactly why I chose to override instead.

> In regard of the rules it is a bit more difficult and I am still not sure about that. I want to let them be included and enabled by default, as locking yourself out from ssh for example is not the first expirience I want new users to have.
> [...]
> What do you think about this? I think the local zone part is pretty uncontroversial, but the rules part might not be.

I generally agree with your assessment. An option for the local zone name is perfect. I also believe that a complex interface for enabling rules is not a good experience, so a common prefix sounds like a good solution to me. I was originally suggesting an opt-in version for the rules because nixos generally uses `enableRecommendedXYZ` as opt-ins also in other modules. This is the part that I'd change personally, because to me it does feel cleaner when I have to opt-in to opinionated defaults. It kind-of suggests that I won't be surprised with new opt-out defaults at a later point. Nonetheless, an opt-out would of course also be a working and fine solution from a technical point of view. Changing this to an opt-in now is still possible in my opinion, but might not be at a later point.

You make a fair point regarding first user experience, although I think a simple usage example that includes `enableRecommendedRules` in the main README with a short comment would also address this. I don't think there's an introduction or usage example in the README currently, so whoever uses this right now probably has to know how it works. Only slightly related, having to manually setup a `stopRuleset` when using this module is probably the bigger hurdle for anyone unfamiliar with nftables.

In conclusion I agree with all your points and the suggested solutions but personally would like to see an opt-in for the rules instead of an opt-out for the reasons stated above. Either way, thanks a lot for your consideration and time!

**Comment by thelegy** [comment]
Hi there. I was able to discuss the topic (with your remarks) with a friend of mine and got some new perspective. I think I have a plan in my mind and will start implementing it in the next couple of days.

For the ruleset I have decided to add a prefix for the names of the rules and move all the rules in another module. This module will expose options that are even more high-level and can add rules. I will add some kind of enable option in there, but some other options will follow as the rules are currently *very* opinionated.

Let me take ICMP as an example. It is efficient to only match the traffic once and then have a list of all types to allow. But fo example currently it always allows `ping` and it does not handle multicast traffic. I would like to add options, to make this configurable. This would probably include high-level enable options for the builtin rules.

Regarding opt-in vs opt-out: I am leaning to let the new high-level enable-option also enable the current one, so the README can point to the high-level option to enable for convenience. That means, that such an enable option would need to be opt-in indeed.

I will need to start implementing and actually see some code to figure out the details, like the names of optiions / the builtin rule prefix.

**Comment by thelegy** [comment]
I started to work on this issue a couple of months ago, but then reallife happened and I had very little time in between. Now I have added most of the things, that were already outlined in this issue. I will still improve the documentation a little, but since the implementation is done, I have already opened an PR for people to look at.

Maybe, if you have time, you can have a look at the PR. Just please don't feel any pressure! I intend to merge it in a few weeks anyways, if I don't hear negative feedback.

**Comment by oddlama** [comment]
Looks great so far, thanks for all your work! I've updated my config to use the draft and on first look everything seems to work as expected :) I'll let you know in case I notice anything else.
